src/main/java/cs361/battleships/models/Board.java

```diff
@@ -74,12 +79,58 @@ public boolean placeShip(Ship ship, int x, char y, boolean isVertical) {
 		return true;
 	}
 
-	/*
-	DO NOT change the signature of this method. It is used by the grading scripts.
+	/**
+	 * Computes the result of an attack on a specific square.
+	 * Verified validity of attack and adds it to list of attacks.
+	 *
+	 * @param x set X coordinate of the attack square.
+	 * @param y set Y coordinate of the attack square.
+	 * @return Result the outcome of the attack.
 	 */
 	public Result attack(int x, char y) {
-		//TODO Implement
-		return null;
+	    Result outcome = new Result();
+	    outcome.setLocation(new Square(x, y));
+
+		if (x < 1 || x > 10 || y < 'A' || y > 'J') {
+            outcome.setResult(INVALID);
+            return outcome;
+		}
+        for (Ship s : ships) {
+			if (s.getOccupiedSquares().contains(outcome.getLocation())) {
+				outcome.setShip(s);
+				outcome.setResult(HIT);
+				break;
+			}
+		}
+        if (outcome.getResult() != HIT) {
+			outcome.setResult(MISS);
+			attacks.add(outcome);
+			return outcome;
+		}
+
+		int hitCount = 0;
+		for (Result r : attacks) {
+			if (r.getLocation().equals(outcome.getLocation())) {
+				outcome.setResult(INVALID);
+				return outcome;
+			} else if (r.getShip().getShipName().equals(outcome.getShip().getShipName())) {
```

Comment:
On my end, if the `attacks` contain any `MISS`, when it runs to this line, the `null` ship will cause `NullPointerException`.

Reply:
If you try to test this case, modify the 46 line of the `Game` class:
from
```java
opponentAttackResult.getResult() != INVALID
```
to
```java
opponentAttackResult.getResult() == INVALID
```
then the AI bug will not obstruct you.

Reply:
Great catch! Bug should be fixed now.